File: src/main/java/de/eventmodelers/support/internal/debug/EventsDebugController.java
package de.eventmodelers.support.internal.debug;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.StreamSupport;
import org.axonframework.config.Configuration;
import org.axonframework.eventhandling.DomainEventMessage;
import org.axonframework.eventsourcing.eventstore.EventStorageEngine;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.web.bind.annotation.*;

enum ElementType {
  EVENT,
  DEAD_LETTER
}

record FieldInfo(String name, String type, String example) {}

record EventMessageInfo(
    String id,
    String title,
    List<FieldInfo> fields,
    ElementType type,
    String description,
    boolean highlighted) {

  EventMessageInfo(
      String id, String title, List<FieldInfo> fields, ElementType type, String description) {
    this(id, title, fields, type, description, false);
  }
}

record EventsResponse(List<EventMessageInfo> events) {}

@ConditionalOnProperty(
    name = "application.debug.enabled",
    havingValue = "true",
    matchIfMissing = false)
@RestController
public class EventsDebugController {

  private final EventStorageEngine eventStorageEngine;
  private final Configuration configuration;

  public EventsDebugController(EventStorageEngine eventStorageEngine, Configuration configuration) {
    this.eventStorageEngine = eventStorageEngine;
    this.configuration = configuration;
  }

  @CrossOrigin
  @GetMapping("/internal/debug/events/{aggregateId}")
  public List<? extends DomainEventMessage<?>> resolveEvents(
      @PathVariable("aggregateId") String aggregateId) {
    return eventStorageEngine.readEvents(aggregateId).asStream().toList();
  }

  @CrossOrigin
  @GetMapping("/internal/debug/history/{aggregateId}")
  public EventsResponse resolveHistory(
      @RequestParam(value = "processingGroup", required = false) String processingGroup,
      @PathVariable("aggregateId") String aggregateId) {

    var domainEvents =
        StreamSupport.stream(
                eventStorageEngine.readEvents(aggregateId).asStream().spliterator(), false)
            .map(e -> (DomainEventMessage<?>) e)
            .toList();

    var mappedDomainEvents =
        domainEvents.stream()
            .map(
                event ->
                    new EventMessageInfo(
                        event.getIdentifier(),
                        camelCaseToWords(event.getPayloadType().getSimpleName()),
                        extractFields(event.getPayload()),
                        ElementType.EVENT,
                        ""))
            .toList();

    if (processingGroup != null) {
      var dlq =
          configuration
              .eventProcessingConfiguration()
              .deadLetterQueue(processingGroup)
              .orElse(null);
      if (dlq != null) {
        var letters = new ArrayList<EventMessageInfo>();
        for (var sequence : dlq.deadLetters()) {
          for (var letter : sequence) {
            if (letter.message() instanceof DomainEventMessage<?> msg) {
              var cause = letter.cause().orElse(null);
              letters.add(
                  new EventMessageInfo(
                      msg.getAggregateIdentifier(),
                      msg.getPayloadType().getSimpleName(),
                      List.of(),
                      ElementType.EVENT,
                      cause != null ? cause.message() : "",
                      true));
            }
          }
        }
        var combined = new ArrayList<>(mappedDomainEvents);
        combined.addAll(letters);
        return new EventsResponse(combined);
      }
    }
    return new EventsResponse(mappedDomainEvents);
  }

  private String camelCaseToWords(String str) {
    return str.replaceAll("([a-z])([A-Z])", "$1 $2")
        .replaceAll("([A-Z]+)([A-Z][a-z])", "$1 $2")
        .trim();
  }

  private List<FieldInfo> extractFields(Object payload) {
    var fields = new ArrayList<FieldInfo>();
    for (Field field : payload.getClass().getDeclaredFields()) {
      field.setAccessible(true);
      try {
        var value = field.get(payload);
        fields.add(
            new FieldInfo(
                field.getName(),
                field.getType().getSimpleName(),
                value != null ? value.toString() : ""));
      } catch (Exception e) {
        // skip inaccessible fields
      }
    }
    return fields;
  }
}

File: src/main/kotlin/de/eventmodelers/support/internal/debug/EventsDebugController.kt
package de.eventmodelers.support.internal.debug

import org.axonframework.config.Configuration
import org.axonframework.eventhandling.DomainEventMessage
import org.axonframework.eventsourcing.eventstore.EventStorageEngine
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

enum class ElementType {
  EVENT,
  DEAD_LETTER
}

data class FieldInfo(val name: String, val type: String, val example: String)

data class EventMessageInfo(
    val id: String,
    val title: String,
    val fields: List<FieldInfo>,
    val type: ElementType,
    val description: String,
    val highlighted: Boolean = false
)

data class EventsResponse(val events: List<EventMessageInfo>)

@ConditionalOnProperty("application.debug.enabled", havingValue = "true", matchIfMissing = false)
@RestController
class EventsDebugController(
    val eventStorageEngine: EventStorageEngine,
    val configuration: Configuration
) {

  @CrossOrigin
  @GetMapping("/internal/debug/events/{aggregateId}")
  fun resolveEvents(@PathVariable("aggregateId") aggregateId: String): List<DomainEventMessage<*>> {
    return eventStorageEngine.readEvents(aggregateId.toString()).asSequence().toList()
  }

  @CrossOrigin
  @GetMapping("/internal/debug/history/{aggregateId}")
  fun resolveHistory(
      @RequestParam("processingGroup", required = false) processingGroup: String?,
      @PathVariable("aggregateId") aggregateId: String
  ): EventsResponse {
    val domainEvents = eventStorageEngine.readEvents(aggregateId).asSequence().toList()

    val mappedDomainEvents =
        domainEvents.map { event ->
          EventMessageInfo(
              id = event.identifier,
              title = camelCaseToWords(event.payloadType.simpleName),
              fields = extractFields(event.payload),
              type = ElementType.EVENT,
              description = "")
        }

    if (processingGroup != null) {
      val dlq =
          configuration.eventProcessingConfiguration().deadLetterQueue(processingGroup).orElse(null)
      var letters: List<EventMessageInfo> =
          dlq.deadLetters()
              .toList()
              .flatMap { it.asSequence() }
              .filter { it.message() is DomainEventMessage<*> }
              .map { Pair(it.message() as DomainEventMessage<*>, it.cause().orElse(null)) }
              .map {
                EventMessageInfo(
                    it.first.aggregateIdentifier,
                    it.first.payloadType.simpleName,
                    emptyList(),
                    ElementType.EVENT,
                    it.second?.message() ?: "",
                    highlighted = true)
              }

      return EventsResponse(mappedDomainEvents + letters)
    } else {
      return EventsResponse(mappedDomainEvents)
    }
  }

  private fun camelCaseToWords(str: String): String {
    return str.replace(Regex("([a-z])([A-Z])"), "$1 $2")
        .replace(Regex("([A-Z]+)([A-Z][a-z])"), "$1 $2")
        .trim()
  }

  private fun extractFields(payload: Any): List<FieldInfo> {
    return payload::class.java.declaredFields.mapNotNull { field ->
      field.isAccessible = true
      try {
        FieldInfo(
            name = field.name,
            type = field.type.simpleName,
            example = field.get(payload)?.toString() ?: "")
      } catch (e: Exception) {
        null
      }
    }
  }

  private fun extractFieldsFromMap(payload: Map<String, Any?>): List<FieldInfo> {
    return payload.map { (key, value) ->
      FieldInfo(name = key, type = "String", example = value?.toString() ?: "")
    }
  }
}
